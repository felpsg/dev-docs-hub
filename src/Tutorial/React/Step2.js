import React from 'react';
import { REACT_CARD_CONTENTS } from '../../pages/ReactDoc';
import '../../assets.js/styles/Step1.css';

function Step2() {
  const content = REACT_CARD_CONTENTS[1]; // Introdução ao React

  return (
    <div className="custom-card">
      <div className="custom-card-header">{content.header}</div>
      <div className="custom-card-body">
        <h5 className="title-color">{content.title}</h5>
        <p className="text">{content.text}</p>
        <h6 className="sub-title">{content.subtitle}</h6>
        <p className="text">{content.text2}</p>
        <h6 className="sub-title">{content.subtitle2}</h6>
        <p className="text">{content.text2}</p>
        <h6 className="sub-title">{content.subtitle3}</h6>
        <p className="text">{content.text3}</p>
        <p className="conclusion">{content.conclusion}</p>
      </div>
    </div>
  );
}

export default Step2;
