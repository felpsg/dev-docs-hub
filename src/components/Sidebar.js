import React, { useState, useEffect, useRef } from 'react';
import { Link, useLocation } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { setStep } from '../store/actions';
import '../assets.js/styles/sidebar.css';

const links = [
  { path: '/home', label: 'Home' },
  { path: '/chat', label: 'HubeGPT' },
  {
    path: '/tutorial/javascript',
    label: 'JavaScript Tutorial',
    subLinks: [
      { path: '/tutorial/javascript/step1', label: 'Inicio' },
      { path: '/tutorial/javascript/step2', label: 'Introdução ao javascript' },
      { path: '/tutorial/javascript/step3', label: 'Vantagens e Desvantagens' },
      { path: '/tutorial/javascript/step4', label: 'Lógica de Programação' },
    ],
  },
  {
    path: '/tutorial/react',
    label: 'React Tutorial',
    subLinks: [
      { path: '/tutorial/react/step1', label: 'Inicio' },
      { path: '/tutorial/react/step2', label: 'Introdução ao React' },
      { path: '/tutorial/react/step3', label: 'Vantagens e Desvantagens' },
    ],
  },
  { path: '/docs/vue', label: 'Vue' },
  { path: '/docs/bootstrap', label: 'Bootstrap' },
  { path: '/docs/tailwind', label: 'Tailwind' },
];

const getStepFromPath = (path) => {
  const match = path.match(/\/tutorial\/(?:javascript|react)\/step(\d+)/);
  return match ? Number(match[1]) : null;
};

const SubMenu = ({ subLinks, location }) => (
  <ul className="sub-menu">
    {subLinks.map((subLink, index) => (
      <li key={index} className={subLink.path === '/tutorial/step1' ? 'special-item' : ''}>
        <Link
          to={subLink.path}
          className={`text-decoration-none ${
            location.pathname === subLink.path ? 'active-link' : ''
          }`}
        >
          {subLink.label}
        </Link>
      </li>
    ))}
  </ul>
);

const Sidebar = () => {
  const [openSubMenu, setOpenSubMenu] = useState(null);
  const openSubMenuRef = useRef(openSubMenu);
  openSubMenuRef.current = openSubMenu;
  const location = useLocation();
  const dispatch = useDispatch();

  const handleSubMenuToggle = (index, hasSubLinks) => {
    if (hasSubLinks) {
      setOpenSubMenu(openSubMenu === index ? null : index);
    }
  };

  const closeSubMenus = () => {
    setOpenSubMenu(null);
  };

  useEffect(() => {
    const stepFromPath = getStepFromPath(location.pathname);
    if (stepFromPath) {
      dispatch(setStep(stepFromPath));
    }

    if (openSubMenuRef.current === null) {
      const matchingLinkIndex = links.findIndex((link) =>
        location.pathname.startsWith(link.path),
      );
      setOpenSubMenu(matchingLinkIndex);
    }
  }, [location.pathname, dispatch]);

  return (
    <div className="col-md-3 sidebar">
      <ul className="list-group">
        {links.map((link, index) => (
          <li className="list-group-item" key={index}>
            {link.subLinks ? (
              <div
                onClick={() => handleSubMenuToggle(index, !!link.subLinks)}
                className={`text-decoration-none ${
                  location.pathname.startsWith(link.path) ? 'active-link' : ''
                }`}
              >
                {link.label}
              </div>
            ) : (
              <Link
                to={link.path}
                onClick={closeSubMenus} // Feche os submenus ao clicar em um link principal
                className={`text-decoration-none ${
                  location.pathname.startsWith(link.path) ? 'active-link' : ''
                }`}
              >
                {link.label}
              </Link>
            )}
            {openSubMenu === index && link.subLinks && (
              <SubMenu subLinks={link.subLinks} location={location} />
            )}
          </li>
        ))}
      </ul>
    </div>
  );
};

export default Sidebar;
