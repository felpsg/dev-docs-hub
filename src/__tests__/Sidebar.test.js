import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Sidebar from '../components/Sidebar';

const mockStore = configureStore([]);
let store;

beforeEach(() => {
  store = mockStore({
    tutorial: { step: 1 },
  });
});

describe('Sidebar', () => {
  test('renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Sidebar />
        </Router>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('opens submenu when clicked', () => {
    render(
      <Provider store={store}>
        <Router>
          <Sidebar />
        </Router>
      </Provider>,
    );

    const linkWithSubMenu = screen.getByText('CSS');
    fireEvent.click(linkWithSubMenu);

    expect(screen.getByText('Inicio')).toBeInTheDocument();
    expect(screen.getByText('Introdução ao CSS')).toBeInTheDocument();
    expect(screen.getByText('Vantagens e Desvantagens')).toBeInTheDocument();
  });

  test('closes submenu when submenu link is clicked again', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Sidebar />
        </Router>
      </Provider>,
    );

    // Abra o submenu 'CSS'
    const linkWithSubMenu = screen.getByText('CSS');
    fireEvent.click(linkWithSubMenu);

    // Feche o submenu 'CSS' clicando novamente no link 'CSS'
    fireEvent.click(linkWithSubMenu);

    // Verifique se o submenu 'CSS' foi fechado
    expect(screen.queryByText('Inicio')).not.toBeInTheDocument();
    expect(screen.queryByText('Introdução ao CSS')).not.toBeInTheDocument();
    expect(
      screen.queryByText('Vantagens e Desvantagens'),
    ).not.toBeInTheDocument();

    // Abra o submenu 'JavaScript Tutorial'
    fireEvent.click(screen.getByText('JavaScript Tutorial'));

    // Verifique se o submenu 'JavaScript Tutorial' foi aberto
    expect(screen.getByText('Inicio')).toBeInTheDocument();
    expect(screen.getByText('Introdução ao javascript')).toBeInTheDocument();
    expect(screen.getByText('Vantagens e Desvantagens')).toBeInTheDocument();
    expect(screen.getByText('Lógica de Programação')).toBeInTheDocument();

    // Feche o submenu 'JavaScript Tutorial' clicando novamente no link 'JavaScript Tutorial'
    fireEvent.click(screen.getByText('JavaScript Tutorial'));

    // Verifique se o submenu 'JavaScript Tutorial' foi fechado
    expect(screen.queryByText('Inicio')).not.toBeInTheDocument();
    expect(
      screen.queryByText('Introdução ao javascript'),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText('Vantagens e Desvantagens'),
    ).not.toBeInTheDocument();
    expect(screen.queryByText('Lógica de Programação')).not.toBeInTheDocument();
  });
});