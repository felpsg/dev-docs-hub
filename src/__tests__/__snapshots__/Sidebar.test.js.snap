// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders correctly 1`] = `
<DocumentFragment>
  <div
    class="col-md-3 sidebar"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item"
      >
        <a
          class="text-decoration-none "
          href="/home"
        >
          Home
        </a>
      </li>
      <li
        class="list-group-item"
      >
        <a
          class="text-decoration-none "
          href="/chat"
        >
          HubeGPT
        </a>
      </li>
      <li
        class="list-group-item"
      >
        <a
          class="text-decoration-none "
          href="/tutorial/vue"
        >
          Vue
        </a>
      </li>
      <li
        class="list-group-item"
      >
        <div
          class="text-decoration-none "
        >
          CSS
        </div>
      </li>
      <li
        class="list-group-item"
      >
        <div
          class="text-decoration-none "
        >
          JavaScript Tutorial
        </div>
      </li>
      <li
        class="list-group-item"
      >
        <div
          class="text-decoration-none "
        >
          React Tutorial
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
